### PR TITLE
Migrating from setup.py install/develop to leverage pip standards

### PR DESCRIPTION
**Pull request by samypr100**
This PR is an attempt to migrate away from `python setup.py install` and `python setup.py develop` into a standards driven install, namely `pip install .` or `pip install -e .`.

These particular command variants have been deprecated for quite some time and could eventually be removed by setuptools.

* https://packaging.python.org/en/latest/discussions/setup-py-deprecated
* https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

The most notable issues of using these deprecated commands is that egg builds (.egg-info) are produced when using them. Eggs have been deprecated https://peps.python.org/pep-0715/ and is notably causing issues with modern build tooling such as `uv` from being able to detect the pytorch package in docker images (e.g. `pytorch/pytorch:2.2.2`) as discussed in https://github.com/astral-sh/uv/issues/2928

`setup.py` command arguments are also deprecated since they're not compatible with a standards driven install. When used for `install/develop`. For now, they were migrated to use [pip config-settings](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-C) which setuptools supports via their  [ConfigSettingsTranslator](https://github.com/pypa/setuptools/blob/main/setuptools/build_meta.py#L159) for a couple of years.

I decided not to modify `sdist/bdist_wheel` and other variants to keep the scope of the PR small, although this is a good start to eventually start migrating away.

Open to feedback in case I missed an important change somewhere.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

**Comment by samypr100** [comment]
@zou3519 Would you be able to re-approve the runs? I've fixed the linting error and rebased from main. Thanks

File: benchmarks/operator_benchmark/README.md [hunk 1]
```diff
@@ -20,7 +20,7 @@ Key Features:
 The instruction below installs a cpp\_extension for PyTorch and it is required to run the benchmark suite.
 ```bash
 cd pt_extension
-python setup.py install
+python -m pip install .
```
Comment (malfet) [inline benchmarks/operator_benchmark/README.md:23]: Had somebody tested that it works?

File: torch/utils/benchmark/examples/prepare_e2e.sh [hunk 2]
```diff
@@ -35,32 +35,32 @@ wget https://github.com/pytorch/pytorch/pull/39744.diff
...
-python setup.py install
+python -m pip install .
 conda deactivate
 
 source activate pr_39744
 git apply 39744.diff
-python setup.py install
+python -m pip install .
 git checkout .
 conda deactivate
```
Comment (malfet) [inline torch/utils/benchmark/examples/prepare_e2e.sh:66]: I'm not sure about the status of those scripts. Do they work right now? If not, perhaps let's leave it as is

**Review by malfet — CHANGES_REQUESTED** [review]
Do you mind splitting this PR into several smaller ones with a proper test plans?
I.e. if CI is passing than all scripts touched by CI are good to go, but I don't want to accept changes to say `utils/benchmark/examples` as I'm not sure they'll still function as expected after the change

Also, just recently I had to deal with an issue (cc @msaroufim ) in https://github.com/pytorch-labs/ao/pull/135 when changing an installation command resulted in pip creating some sort of isolated environment where it could not find required dependencies as `setup.py` is written in a way that it expects build time dependencies available by default and they are not mentioned in `pyproject.toml`

I know this might sounds a bit unusual, but if one is building PyTorch in say conda environment (or using python runtime from Ubuntu), one might have `setuptools` package available, but not `pip`. Which makes me wonder, why we should force people to install `pip` if it might not be available on their system

**Comment by samypr100** [comment]
Closing for now since the upstream `uv` issue has been resolved, appreciate your patience during the review!
If anyone with more suitable hardware to test all the required build configurations wants to pick this up again that would be awesome 👍

**Pull request closed without merging**